File: lib/python/openlinktoken-core-ai/src/main/openlinktoken/core/ai/tokentransformer/rotation/embedding_transformer.py
"""Protocol for embedding transformers that convert float vectors to token strings."""

from typing import List, Protocol, runtime_checkable


@runtime_checkable
class EmbeddingTransformer(Protocol):
    """Protocol for transformers that convert a float embedding to a list of token strings."""

    def transform(self, embedding: List[float]) -> List[str]:
        """Transform a raw float embedding into a list of token strings."""
        ...

File: lib/java/openlinktoken-core-ai/src/main/java/org/openlinktoken/core/ai/tokentransformer/rotation/EmbeddingTransformer.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.core.ai.tokentransformer.rotation;

import java.util.List;

/**
 * Transforms a raw float embedding vector into a list of token strings.
 *
 * <p>Implementations may apply projection, quantization, or other
 * dimensionality-reduction steps to convert a high-dimensional embedding into
 * one or more compact token representations.
 */
public interface EmbeddingTransformer {

    /**
     * Transform an embedding vector into a list of token strings.
     *
     * @param embedding raw float embedding vector
     * @return list of token strings derived from the embedding; never {@code null}
     */
    List<String> transform(float[] embedding);
}
